Fix GHC not building with musl

nh2 opened the pull request:
###### Motivation for this change

Fixes binary GHCs segfaulting on `pkgsMusl`. See #129247, #118731.

All Nix-built Haskell compilers build now on `pkgsMusl`:

```sh
NIX_PATH=nixpkgs=. nix-build --no-link --expr 'with import <nixpkgs> {}; { inherit (pkgsMusl.haskell.compiler) ghc884 ghc8104 ghc901 ghcHEAD; }'
```

Key changes included (see commits for details):

* Switch `ghc8102Binary` to use GHC HQ musl bindist instead of patching glibc bindist with musl.
* Use this better binary bindist to bootstrap all nix-built GHCs for `pkgsMusl`
* Add missing `hardeningDisable = [ "pie" ]` where needed.
* Swith to build GHC without sphinx by default for musl (cross already does this) to avoid huge amount of build dependencies, some of which may not work with musl yet (or break frequently). This will reduce maintenance burden for projects like [`static-haskell-nix`](https://github.com/nh2/static-haskell-nix/) significantly, and make it cheaper to build musl-based GHC on Hydra.
  * Add new options to control this.
* Add new structure to declare which `ncurses` version and similar dependencies the bindists have, and a `postUnpack` step to ensure at build time that those are actually correct.
  * The build now complains when your expectation is wrong regarding what `ncurses` a bindist needs.
* Fix i686 build of `ghc8102Binary` as a drive-by contribution.
* Lots of explanatory extra comments.

###### Extra review needed for

* My usage of `buildPlatform`, `targetPlatform` etc.
* There are some some cross options that I slightly touch. People who care about cross Haskell should ideally check out if everything remains working.

###### TODO

* [x] I need some help to make the Hydra evaluation correct, see top `WIP` commit.
  * provided by @sternenseemann in https://github.com/NixOS/nixpkgs/pull/129289#pullrequestreview-699418026

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Relase notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

CC people who care about this and may want to try whether it breaks anything, and others qualified to review:

@NixOS/haskell @expipiplus1 @peti @bgamari @nomeata @domenkozar @cdepillabout @dtzWill @lunaris @NorfairKing @maralorn @sternenseemann @dtzWill @Ericson2314 @matthewbauer @aveltras

nh2 committed:
haskell.packages.ghc8102Binary: Add comments

nh2 committed:
ghc: Refactor: Extract `ghcBinDists`

nh2 committed:
haskell.compiler.ghc8102Binary: Fix build on i686.

The library override that was present in the code referred to a
name that isn't even used in current GHC bindists.

Tested with:

    NIX_PATH=nixpkgs=. nix-build --no-link -A haskell.compiler.ghc8102Binary --argstr system i686-linux

nh2 committed:
haskell.compiler.ghc8102Binary: Add check for arch-specific libraries.

With this check, we no longer don't notice when the upstream bindist
changes its dependencies (e.g. because a newer Debian version is used
that uses a new `ncurses` version).

nh2 committed:
haskell.compiler.ghc8102Binary: Remove `sed` of `/usr/bin/*` commands.

None of the current bindists appear to contain these paths in their
`ghc-stage2` binary.

nh2 committed:
pkgsMusl.haskell.compiler.ghc8102Binary: Fix musl segfault. Fixes #118731 #129247.

This commit replaces the musl + glibc hackery in the GHC bindist
compiler by using the new musl based bindist that GHC HQ provides
(built on Alpine).
We could alternatively also use a nix-built musl boostrap compiler,
but it seems nicer to use the GHC HQ one for now.

This fixes the compiler built by
`pkgsMusl.haskell.compiler.ghc8102Binary` segfaulting (#118731)
since the commit

    5e2311d2f - musl: 1.2.1 -> 1.2.2

concretely, musl commit

    01c7920f - remove redundant pthread struct members repeated for layout purposes

which I suspect breaks some glibc/musl ABI compatibility that may have
existed accidentally until then.

The added

    lib.optional stdenv.targetPlatform.isMusl "pie";

also fixes that the packaged bindist compiler cannot create a binary
in its `installCheck` phase (and overall); see detail explanation
in #129247.

nh2 committed:
haskell.compiler.ghc8102Binary: find `-executable` -> `-perm -0100`

nh2 committed:
haskell.compiler.ghc865Binary: Disable "pie" hardening. See #129247

While this does not fix `ghc865Binary` with musl, it at least prevents
that the other, newer errors are shadowed (see #129247).

nh2 committed:
pkgsMuslhaskell.compiler.{ghc884,ghc8104}: Use GHC 8.10 as bootstrap compiler.

This addresses the fact that `ghc865Binary` segfaults on musl
(see #118731) because of the glibc+musl mix used in there.

With the previous commits, `ghc8102Binary` was changed to use
the musl-based bindist from GHC HQ instead, which works.

With this change, all nix Haskell compilers builds on musl:

    NIX_PATH=nixpkgs=. nix-build --no-link --expr 'with import <nixpkgs> {}; { inherit (pkgsMusl.haskell.compiler) ghc884 ghc8104 ghc901 ghcHEAD; }'

nh2 committed:
pkgsMusl.haskell.compiler.ghc901: Fix evaluation with musl

`glibcLocales` only exists when glibc is used.

Similar to commit:

    8727284a - haskell: only use glibcLocales when using glibc

nh2 committed:
ghc: Add comments about `hardeningDisable` `pie` for musl

nh2 committed:
pkgsMusl.haskell.compiler.ghc{8104,884,901,HEAD}: Disable sphinx for musl

Adds new package options:

* enableDocs
* enableHaddockProgram

to control whether to build Sphinx docs, and GHC haddocks and the
haddock program.

Unfortunately currently the building of the `haddock `program
and generating GHC docs are mixed into one option, see:
https://gitlab.haskell.org/ghc/ghc/-/issues/20077

Making Sphinx docs disableable, and disabling them by default
for Musl and cross builds, makes it much easier to provide these
builds without having to support Sphinx's enormous dependency
tree for those ways of building.

nh2 committed:
argp-standalone: Add `-fPIC`.

Fixes `pkgsMusl.elfutils` failing with `recompile with -fPIC`.

This was discovered trying to build `pkgsMusl.haskell.compiler.ghcHEAD`.

nh2 committed:
release-haskell.nix: Add pkgsMusl haskell compilers to Hydra jobs

nh2 commented:
Looks like I broke the nixpkgs manual with my new doc-related config options:

```
ghc.mk:184: *** BUILD_SPHINX_PDF=YES, but `xelatex` was not found. Install `xelatex`, then rerun `./configure`.
```

https://github.com/NixOS/nixpkgs/runs/2986064008?check_suite_focus=true#step:5:3174

Not sure why yet.

nh2 commented:
Will need to target `staging` as this is a mass rebuild (probably via  `pandoc`? I haven't investigated yet).

maralorn commented:
Well pandoc is not really the point here. I believe we use the 8102 binary to build our 8104 ghc? So this is a full haskellPackages rebuild. We normally do those on haskell-updates, so I don‘t think this needs to target staging, but haskell-updates.

r-rmcgibbo commented:
Result of `nixpkgs-review pr 129289` at a84215ba run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>29 packages marked as broken and skipped:</summary>
  <ul>
    <li>grass</li>
    <li>libvmi</li>
    <li>mvapich</li>
    <li>python38Packages.dask-gateway</li>
    <li>python38Packages.dask-jobqueue</li>
    <li>python38Packages.dask-ml</li>
    <li>python38Packages.dask-mpi</li>
    <li>python38Packages.dask-xgboost</li>
    <li>python38Packages.datashader</li>
    <li>python38Packages.fipy</li>
    <li>python38Packages.pygmt</li>
    <li>python38Packages.streamz</li>
    <li>python38Packages.stumpy</li>
    <li>python38Packages.tensorflowWithCuda</li>
    <li>python39Packages.dask-gateway</li>
    <li>python39Packages.dask-jobqueue</li>
    <li>python39Packages.dask-mpi</li>
    <li>python39Packages.dask-xgboost</li>
    <li>python39Packages.fipy</li>
    <li>python39Packages.pygmt</li>
    <li>python39Packages.streamz</li>
    <li>python39Packages.tensorflowWithCuda</li>
    <li>qemu_xen</li>
    <li>qemu_xen-light</li>
    <li>qemu_xen_4_10</li>
    <li>qemu_xen_4_10-light</li>
    <li>qlandkartegt</li>
    <li>qubes-core-vchan-xen</li>
    <li>yosys-bluespec</li>
  </ul>
</details>
<details>
  <summary>6987 packages skipped due to time constraints:</summary>
  <ul>
    <li>EBTKS</li>
    <li>agdaPackages.agda-categories</li>
    <li>agdaPackages.agda-prelude</li>
    <li>agdaPackages.cubical</li>
    <li>agdaPackages.functional-linear-algebra</li>
    <li>agdaPackages.generic</li>
    <li>agdaPackages.iowa-stdlib</li>
    <li>agdaPackages.standard-library</li>
    <li>apacheHttpdPackages.mod_tile (apacheHttpdPackages_2_4.mod_tile)</li>
    <li>apostrophe</li>
    <li>...</li>
  </ul>
</details>
<details>
  <summary>2 packages built successfully:</summary>
  <ul>
    <li>argp-standalone</li>
    <li>haskell.compiler.ghc8102BinaryMinimal</li>
  </ul>
</details>
<details>
  <summary>5 suggestions:</summary>
  <ul>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/missing-patch-comment.md">missing-patch-comment</a>

Consider adding a comment explaining the purpose of this patch on the line preceeding.
Near pkgs/development/libraries/argp-standalone/default.nix:31:53:
```
   |
31 |        lib.optionals stdenv.hostPlatform.isDarwin [ patch-argp-fmtstream ]
   |                                                     ^
```
</li>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/missing-patch-comment.md">missing-patch-comment</a>

Consider adding a comment explaining the purpose of this patch on the line preceeding.
Near pkgs/development/libraries/argp-standalone/default.nix:32:52:
```
   |
32 |     ++ lib.optionals stdenv.hostPlatform.isLinux [ patch-throw-in-funcdef patch-shared ];
   |                                                    ^
```
</li>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/name-and-version.md">name-and-version</a>

Did you mean to pass `pname` instead of `name` to `mkDerivation`?

Near pkgs/development/compilers/ghc/8.10.2-binary.nix:135:3:
```
    |
135 |   name = "ghc-${version}-binary";
    |   ^
```

Near pkgs/development/compilers/ghc/8.10.2-binary.nix:133:10:
```
    |
133 |   inherit version;
    |          ^
```
</li>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/unclear-gpl.md">unclear-gpl</a>

`gpl2` is a deprecated license, please check if project uses `gpl2Plus` or `gpl2Only` and change `meta.license` accordingly.

Near pkgs/development/libraries/argp-standalone/default.nix:59:5:
```
   |
59 |     license = licenses.gpl2;
   |     ^
```
</li>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/missing-patch-comment.md">missing-patch-comment</a>

Consider adding a comment explaining the purpose of this patch on the line preceeding.
Near pkgs/development/libraries/argp-standalone/default.nix:32:75:
```
   |
32 |     ++ lib.optionals stdenv.hostPlatform.isLinux [ patch-throw-in-funcdef patch-shared ];
   |                                                                           ^
```
</li>
  </ul>
</details>

nh2 commented:
> I don‘t think this needs to target staging, but haskell-updates.

Done.

sternenseemann commented on pkgs/top-level/release-haskell.nix:
```diff
@@ -96,6 +96,16 @@ let
         writers = testPlatforms.writers;
       };
 
+      # TODO: Currently gets me
```
This is due to the throw in `pkgs/top-level/stage.nix` which is really annoying. I'll try to think of a solution, maybe that one should be removed altogether…

sternenseemann replied:
One possibility would be something like this which is a bit ugly as the platform list then becomes eval platform specific. However no platform uses musl by default so far, so the impact of this would be low in practice:

```diff
diff --git a/pkgs/development/compilers/ghc/8.10.2-binary.nix b/pkgs/development/compilers/ghc/8.10.2-binary.nix
index c5c7d935e00..aa9135d9e0a 100644
--- a/pkgs/development/compilers/ghc/8.10.2-binary.nix
+++ b/pkgs/development/compilers/ghc/8.10.2-binary.nix
@@ -36,81 +36,84 @@ let
   # * To skip file checking for a specific arch specfic library,
   #   set `fileToCheckFor = null`.
   ghcBinDists = {
-    i686-linux = {
-      src = {
-        url = "${downloadsUrl}/${version}/ghc-${version}-i386-deb9-linux.tar.xz";
-        sha256 = "0bvwisl4w0z5z8z0da10m9sv0mhm9na2qm43qxr8zl23mn32mblx";
+    # binary distributions for the default libc nixpkgs uses for the respective system
+    default = {
+      i686-linux = {
+        src = {
+          url = "${downloadsUrl}/${version}/ghc-${version}-i386-deb9-linux.tar.xz";
+          sha256 = "0bvwisl4w0z5z8z0da10m9sv0mhm9na2qm43qxr8zl23mn32mblx";
+        };
+        exePathForLibraryCheck = "ghc/stage2/build/tmp/ghc-stage2";
+        archSpecificLibraries = [
+          # The i686-linux bindist provided by GHC HQ is currently built on Debian 9,
+          # which link it against `libtinfo.so.5` (ncurses 5).
+          # Other bindists are linked `libtinfo.so.6` (ncurses 6).
+          { nixPackage = ncurses5; fileToCheckFor = "libtinfo.so.5"; }
+        ];
       };
-      exePathForLibraryCheck = "ghc/stage2/build/tmp/ghc-stage2";
-      archSpecificLibraries = [
-        # The i686-linux bindist provided by GHC HQ is currently built on Debian 9,
-        # which link it against `libtinfo.so.5` (ncurses 5).
-        # Other bindists are linked `libtinfo.so.6` (ncurses 6).
-        { nixPackage = ncurses5; fileToCheckFor = "libtinfo.so.5"; }
-      ];
-    };
-    x86_64-linux =
-      if stdenv.hostPlatform.isMusl
-        then
-          { # musl-bindist (referred to below)
-            src = {
-              url = "${downloadsUrl}/${version}/ghc-${version}-x86_64-alpine3.10-linux-integer-simple.tar.xz";
-              sha256 = "0xpcbyaxqyhbl6f0i3s4rp2jm67nqpkfh2qlbj3i2fiaix89ml0l";
-            };
-            exePathForLibraryCheck = "bin/ghc";
-            archSpecificLibraries = [
-              # In contrast to glibc builds, the musl-bindist uses `libncursesw.so.*`
-              # instead of `libtinfo.so.*.`
-              { nixPackage = ncurses6; fileToCheckFor = "libncursesw.so.6"; }
-            ];
-          }
-        else
-          { # normal glibc based bindist
-            src = {
-              url = "${downloadsUrl}/${version}/ghc-${version}-x86_64-deb10-linux.tar.xz";
-              sha256 = "0chnzy9j23b2wa8clx5arwz8wnjfxyjmz9qkj548z14cqf13slcl";
-            };
-            exePathForLibraryCheck = "ghc/stage2/build/tmp/ghc-stage2";
-            archSpecificLibraries = [
-              { nixPackage = ncurses6; fileToCheckFor = "libtinfo.so.6"; }
-            ];
-          }
-        ;
-    armv7l-linux = {
-      src = {
-        url = "${downloadsUrl}/${version}/ghc-${version}-armv7-deb10-linux.tar.xz";
-        sha256 = "1j41cq5d3rmlgz7hzw8f908fs79gc5mn3q5wz277lk8zdf19g75v";
+      x86_64-linux = {
+        src = {
+          url = "${downloadsUrl}/${version}/ghc-${version}-x86_64-deb10-linux.tar.xz";
+          sha256 = "0chnzy9j23b2wa8clx5arwz8wnjfxyjmz9qkj548z14cqf13slcl";
+        };
+        exePathForLibraryCheck = "ghc/stage2/build/tmp/ghc-stage2";
+        archSpecificLibraries = [
+          { nixPackage = ncurses6; fileToCheckFor = "libtinfo.so.6"; }
+        ];
       };
-      exePathForLibraryCheck = "ghc/stage2/build/tmp/ghc-stage2";
-      archSpecificLibraries = [
-        { nixPackage = ncurses6; fileToCheckFor = "libtinfo.so.6"; }
-      ];
-    };
-    aarch64-linux = {
-      src = {
-        url = "${downloadsUrl}/${version}/ghc-${version}-aarch64-deb10-linux.tar.xz";
-        sha256 = "14smwl3741ixnbgi0l51a7kh7xjkiannfqx15b72svky0y4l3wjw";
+      armv7l-linux = {
+        src = {
+          url = "${downloadsUrl}/${version}/ghc-${version}-armv7-deb10-linux.tar.xz";
+          sha256 = "1j41cq5d3rmlgz7hzw8f908fs79gc5mn3q5wz277lk8zdf19g75v";
+        };
+        exePathForLibraryCheck = "ghc/stage2/build/tmp/ghc-stage2";
+        archSpecificLibraries = [
+          { nixPackage = ncurses6; fileToCheckFor = "libtinfo.so.6"; }
+        ];
+      };
+      aarch64-linux = {
+        src = {
+          url = "${downloadsUrl}/${version}/ghc-${version}-aarch64-deb10-linux.tar.xz";
+          sha256 = "14smwl3741ixnbgi0l51a7kh7xjkiannfqx15b72svky0y4l3wjw";
+        };
+        exePathForLibraryCheck = "ghc/stage2/build/tmp/ghc-stage2";
+        archSpecificLibraries = [
+          { nixPackage = ncurses6; fileToCheckFor = "libtinfo.so.6"; }
+          { nixPackage = numactl; fileToCheckFor = null; }
+        ];
+      };
+      x86_64-darwin = {
+        src = {
+          url = "${downloadsUrl}/${version}/ghc-${version}-x86_64-apple-darwin.tar.xz";
+          sha256 = "1hngyq14l4f950hzhh2d204ca2gfc98pc9xdasxihzqd1jq75dzd";
+        };
+        exePathForLibraryCheck = null; # we don't have a library check for darwin yet
+        archSpecificLibraries = [
+          { nixPackage = ncurses6; fileToCheckFor = null; }
+          { nixPackage = libiconv; fileToCheckFor = null; }
+        ];
       };
-      exePathForLibraryCheck = "ghc/stage2/build/tmp/ghc-stage2";
-      archSpecificLibraries = [
-        { nixPackage = ncurses6; fileToCheckFor = "libtinfo.so.6"; }
-        { nixPackage = numactl; fileToCheckFor = null; }
-      ];
     };
-    x86_64-darwin = {
-      src = {
-        url = "${downloadsUrl}/${version}/ghc-${version}-x86_64-apple-darwin.tar.xz";
-        sha256 = "1hngyq14l4f950hzhh2d204ca2gfc98pc9xdasxihzqd1jq75dzd";
+    # binary distributions for the musl libc for the respective system
+    musl = {
+      x86_64-linux = {
+        src = {
+          url = "${downloadsUrl}/${version}/ghc-${version}-x86_64-alpine3.10-linux-integer-simple.tar.xz";
+          sha256 = "0xpcbyaxqyhbl6f0i3s4rp2jm67nqpkfh2qlbj3i2fiaix89ml0l";
+        };
+        exePathForLibraryCheck = "bin/ghc";
+        archSpecificLibraries = [
+          # In contrast to glibc builds, the musl-bindist uses `libncursesw.so.*`
+          # instead of `libtinfo.so.*.`
+          { nixPackage = ncurses6; fileToCheckFor = "libncursesw.so.6"; }
+        ];
       };
-      exePathForLibraryCheck = null; # we don't have a library check for darwin yet
-      archSpecificLibraries = [
-        { nixPackage = ncurses6; fileToCheckFor = null; }
-        { nixPackage = libiconv; fileToCheckFor = null; }
-      ];
     };
   };
 
-  binDistUsed = ghcBinDists.${stdenv.hostPlatform.system}
+  distSetName = if stdenv.hostPlatform.isMusl or false then "musl" else "default";
+
+  binDistUsed = ghcBinDists.${distSetName}.${stdenv.hostPlatform.system}
     or (throw "cannot bootstrap GHC on this platform");
 
   useLLVM = !stdenv.targetPlatform.isx86;
@@ -361,7 +364,11 @@ stdenv.mkDerivation rec {
     homepage = "http://haskell.org/ghc";
     description = "The Glasgow Haskell Compiler";
     license = lib.licenses.bsd3;
-    platforms = ["x86_64-linux" "armv7l-linux" "aarch64-linux" "i686-linux" "x86_64-darwin"];
+    # HACK: since we can't encode the libc / abi in platforms, we need
+    # to make the platform list dependent on the evaluation platform
+    # in order to avoid eval errors with musl which supports less
+    # platforms than the default libcs (i. e. glibc / libSystem)
+    platforms = builtins.attrNames ghcBinDists.${distSetName};
     hydraPlatforms = builtins.filter (p: minimal || p != "aarch64-linux") platforms;
     maintainers = with lib.maintainers; [ lostnet ];
   };
diff --git a/pkgs/top-level/release-haskell.nix b/pkgs/top-level/release-haskell.nix
index c382fc1d732..bf26526872b 100644
--- a/pkgs/top-level/release-haskell.nix
+++ b/pkgs/top-level/release-haskell.nix
@@ -96,14 +96,6 @@ let
         writers = testPlatforms.writers;
       };
 
-      # TODO: Currently gets me
-      #           error: Musl libc only supports Linux systems.
-      #       in
-      #           hydra-eval-jobs -I . pkgs/top-level/release-haskell.nix
-      #       because it generates things like:
-      #           pkgsMusl.haskell.compiler.integer-simple.ghc8104.x86_64-darwin
-      #       Not sure how to turn that off.
-      # `pkgsMusl` based compilers
       pkgsMusl.haskell.compiler = packagePlatforms pkgs.pkgsMusl.haskell.compiler;
 
       # test some statically linked packages to catch regressions
@@ -326,11 +318,11 @@ let
             nh2
           ];
         };
-        constituents = [
-          jobs.pkgsMusl.haskell.compiler.ghc8102Binary.x86_64-linux
-          jobs.pkgsMusl.haskell.compiler.ghc884.x86_64-linux
-          jobs.pkgsMusl.haskell.compiler.ghc8104.x86_64-linux
-          jobs.pkgsMusl.haskell.compiler.ghc901.x86_64-linux
+        constituents = accumulateDerivations [
+          jobs.pkgsMusl.haskell.compiler.ghc8102Binary
+          jobs.pkgsMusl.haskell.compiler.ghc884
+          jobs.pkgsMusl.haskell.compiler.ghc8104
+          jobs.pkgsMusl.haskell.compiler.ghc901
         ];
       };
     }
```

Splitting the `ghcBinDists` set into `musl` and `default` is also a bit nicer in my opinion. The remaining error in the hydra eval is from `ghc865Binary` which would need a similar treatment.

nh2 replied:
> ```diff
> stdenv.hostPlatform.isMusl or false
> ```

Is that necessary? Isn't `isMusl` an always-present bool?

nh2 replied:
Done, I've pushed this change.

sternenseemann replied:
> @sternenseemann Is the `or false` necessary? Isn't `isMusl` an always-present bool?

I'm not sure, if it is according to your testing, it is probably always present :)

roberth commented on pkgs/development/compilers/ghc/8.10.4.nix:
```diff
@@ -38,13 +38,31 @@
+    # Docs disabled for musl and cross because it's a large task to keep
+    # all `sphinx` dependencies building in those environments.
+    # `sphinx` pullls in among others:
+    # Ruby, Python, Perl, Rust, OpenGL, Xorg, gtk, LLVM.
```
> OpenGL, Xorg, gtk, LLVM

(FYI) This seems to be mostly because of `librsvg`. Issue https://github.com/NixOS/nixpkgs/issues/126611

Regardless, I don't think we want the other dependencies for cross, so +1

Ericson2314 replied:
Sphinx is only used at build time, so all these should be no problem except for annoying bugs where stuff is target sensative.

nh2 replied:
@Ericson2314 Yes, the problem with it is that a lot of those build-time dependencies often simply have broken builds in `pkgsMusl`.

sternenseemann replied:
`pkgsMusl.buildPackages` uses `glibc` though, so these *should* just work as normally.

nh2 replied:
Sphinx definitely pulls in musl-based builds of the big stuff when put into `nativeBuildInputs`:

`nix-store -q --tree $(NIX_PATH=nixpkgs=. nix-instantiate -A pkgsMusl.haskell.compiler.ghc8104)`

```
+---/nix/store/rwknqkhd87phxq8iw1hz2kh983za44xi-python3.8-sphinx-4.0.2.drv
    +---/nix/store/ck02h2ph9zsscw3cb59brn669hbzynfm-python3.8-sphinxcontrib-serializinghtml-1.1.5.drv
    |   +---/nix/store/bl58hg9v5ni17wl5255n6gn8ryqn0ncr-librsvg-2.50.6.drv
    |   |   +---/nix/store/dqb7wllys8fi4c05kayp7vis7q7sqyv5-rustc-1.52.1.drv
```

`nix-store -q --tree $(NIX_PATH=nixpkgs=. nix-instantiate -A haskell.compiler.ghc8104)`

```
        +---/nix/store/pr7m6hbnn9xhcwmq1mnxcygnm06qkfgk-librsvg-2.50.6.drv
```

Diffing those two librsvg derivations:

`nix-diff /nix/store/pr7m6hbnn9xhcwmq1mnxcygnm06qkfgk-librsvg-2.50.6.drv /nix/store/bl58hg9v5ni17wl5255n6gn8ryqn0ncr-librsvg-2.50.6.drv`

```
        • The input named `bootstrap-stage0-binutils-wrapper-` differs
          - /nix/store/pmvn7qs31f8fwl04w6qdrzb6kqsjizbg-bootstrap-stage0-binutils-wrapper-.drv:{out}
          + /nix/store/fz2w0rj4rvmbny1hrz8hc0hlhjahffrs-bootstrap-stage0-binutils-wrapper-.drv:{out}
          • The builders do not match
              - /nix/store/p4s4jf7aq6v6z9iazll1aiqwb34aqxq9-bootstrap-tools/bin/bash
              + /nix/store/5b7a4ynm23gm375q1fzbsh4sqc806864-bootstrap-tools/bin/bash
          • The set of input names do not match:
              - bootstrap-stage0-glibc-bootstrap
              + bootstrap-stage0-musl-bootstrap
```

`glibc-bootstrap` vs `musl-bootstrap` in here.

nh2 replied:
Resolved on Matrix: https://matrix.to/#/!RbXGJhHMsnQcNIDFWN:nixos.org/$2dtglPPokoMcxS7yQt53iBLFvMkqr-R3yqdYA5hJyVg?via=nixos.org&via=matrix.org&via=orbstheorem.ch

> oh yeah it isn't cross compiled, nevermind
>
> https://github.com/NixOS/nixpkgs/blob/6cc55fd6380dd892f1d41ba7d07efa2a813d38fe/pkgs/top-level/stage.nix#L178-L193

roberth commented on pkgs/development/libraries/argp-standalone/default.nix:
```diff
@@ -33,7 +33,11 @@ stdenv.mkDerivation {
+  # For currently unknown reason, `-fPIC` has to be passed explicitly, otherwise
+  # downstream software like `elfutils` will get `recompile errors like:
+  #     libargp.a(argp-help.o): relocation R_X86_64_PC32 against symbol `program_invocation_short_name' can not be used when making a shared object; recompile with -fPIC
+  # It seems that nixpkgs's on-by-default `-fFPIC` is not in effect here.
```
```suggestion
  # It seems that nixpkgs's on-by-default `-fPIC` is not in effect here.
```

nh2 replied:
Thanks, done.

roberth commented on pkgs/top-level/haskell-packages.nix:
```diff
@@ -57,7 +57,8 @@ in {
+      # Musl bindists do not exist for ghc 8.6.5, so we use 8.10.* for them
+      bootPkgs = if stdenv.isAarch64 || stdenv.targetPlatform.isMusl then
           packages.ghc8102BinaryMinimal
         else
           packages.ghc865Binary;
```
<sub>
Sorry for the noise, github.com ui doesn't let me remove this review comment :(
</sub>

roberth commented on pkgs/top-level/haskell-packages.nix:
```diff
@@ -57,7 +57,8 @@ in {
     ghc884 = callPackage ../development/compilers/ghc/8.8.4.nix {
       # aarch64 ghc865Binary gets SEGVs due to haskell#15449 or similar
-      bootPkgs = if stdenv.isAarch64 then
+      # Musl bindists do not exist for ghc 8.6.5, so we use 8.10.* for them
+      bootPkgs = if stdenv.isAarch64 || stdenv.targetPlatform.isMusl then
           packages.ghc8102BinaryMinimal
         else
           packages.ghc865Binary;
```
I added the conditional for AArch64 to limit rebuilds at the time. Perhaps we could bootstrap everything with 8.10 in or after this PR.

sternenseemann replied:
Sounds good to me, if we don't use GH 8.6.5 we probably should drop it as well since it would be quite behind 8.10.2 in terms of quality after this PR.

nh2 committed:
haskell.compiler.ghc8102Binary: Split `ghcBinDists` set into default and musl.

This allows to implement the "HACK" mentioned in the commit to
build `pkgsMusl` GHCs on Hydra without failing evaluation on Darwin.

Reference of the discussion:
https://github.com/NixOS/nixpkgs/pull/129289#discussion_r663956747

Patch contributed by @sterni.

nh2 committed:
ghc*-binary: Remove no-op `--with-gmp-libraries`

sternenseemann commented on pkgs/development/compilers/ghc/8.10.2-binary.nix:
```diff
@@ -361,7 +365,11 @@ stdenv.mkDerivation rec {
+    # HACK: since we can't encode the libc / abi in platforms, we need
+    # to make the platform list dependent on the evaluation platform
+    # in order to avoid eval errors with musl which supports less
+    # platforms than the default libcs (i. e. glibc / libSystem)
```
Sorry to bother you about this again, but maybe it would make things clearer if we mention here that we are doing this for the benefit of hydra, so `pkgsMusl.haskell.compiler.ghc8102Binary` can be evaluated properly even with darwin in `supportedSystems`.

(Technically only if hydra is not running on darwin)

nh2 replied:
Sure, I'm happy to make the comment as good as possible -- but you understand the Hydra parts much better than me. If you could make a Github suggestion of how to word it, I'll apply it.

sternenseemann replied:
```suggestion
    # platforms than the default libcs (i. e. glibc / libSystem).
    # This is done for the benefit of Hydra, so `packagePlatforms`
    # won't return any platforms that would cause an evaluation
    # failure for `pkgsMusl.haskell.compiler.ghc8102Binary`, as
    # long as the evaluator runs on a platform that supports
    # `pkgsMusl`.
```

nh2 replied:
Done.

nh2 commented:
I've addressed all feedback. 

`static-haskell-nix` is working quite well with these changes: https://github.com/nh2/static-haskell-nix/pull/98

Ready to merge to `haskell-updates` from my side.

sternenseemann commented on pkgs/development/compilers/ghc/8.10.2-binary.nix:
```diff
@@ -14,68 +15,191 @@
+                fi
+
+                echo "Checking that the nix package ${nixPackage} contains ${fileToCheckFor}"
+                if ! test -e "${lib.makeLibraryPath [nixPackage]}/${fileToCheckFor}"; then
```
Nitpick: I feel like the following would be clearer if equivalent.

```suggestion
                if ! test -e "${lib.getLib nixPackage}/lib/${fileToCheckFor}"; then
```

nh2 replied:
Good point, done.

sternenseemann reviewed (approved):
Looks good to me, although I haven't tried building anything with these changes. I'll try building the default compiler and I guess we'll check for extra regressions via the normal `haskell-updates` workflow.
